### PR TITLE
Fix inference for same (but distinct) record types in seq CE

### LINKED ISSUE #14213: Implicit yield and if-then when using records with similarly-named fields causes type inference error
Consider this:

```f#
type A = { X: int }
type B = { X: int }

let _x: (A -> int) list = [
    fun a -> a.X
    if true then fun a -> a.X
]
```

This fails with a type inference error:

image

Adding `yield` works:

```f#
let _x: (A -> int) list = [
    yield fun a -> a.X
    if true then yield fun a -> a.X
]
```

As does type annotations to both `a` parameters:

```f#
let _x: (A -> int) list = [
    fun (a: A) -> a.X
    if true then fun (a: A) -> a.X
]
```

As does removing the `if`:

```f#
let _x: (A -> int) list = [
    fun a -> a.X
    fun a -> a.X
]
```

As a developer with no knowledge of the compiler, I can't see a compelling reason why I get the error. I would expect it to work.

### PR DESCRIPTION
**Pull request by abonie**
Fixes #14213

The issue was, that with implicit yield, we first don't know whether an expression will have type `unit` or some particular expected type, so we type check it without any additional constraints, but that leads to ignoring type annotation and in case of #14213 we infer wrong type.

My solution so far is check the same expression again if we concluded it is not `unit`, this time with a subtype constraint.

### CODE REVIEW

File: src/Compiler/Checking/CheckComputationExpressions.fs [hunk 1]
```diff
@@ -2210,14 +2210,15 @@ let TcSequenceExpression (cenv: cenv) env tpenv comp (overallTy: OverallTy) m =
...
-            if enableImplicitYield then 
+            if enableImplicitYield then
                 let hasTypeUnit, expr, tpenv = TryTcStmt cenv env tpenv comp
                 if hasTypeUnit then 
                     Choice2Of2 expr, tpenv
                 else
                     let genResultTy = NewInferenceType g
                     let mExpr = expr.Range
                     UnifyTypes cenv env mExpr genOuterTy (mkSeqTy cenv.g genResultTy)
+                    let expr, tpenv = TcExprFlex cenv flex true genResultTy env tpenv comp
```
Comment (abonie) [inline src/Compiler/Checking/CheckComputationExpressions.fs:2221]: A bit worried that this is legacy way of doing this, according to the comment here https://github.com/dotnet/fsharp/blob/739acb3a598a5d22c26132e1547afe842b7cfd83/src/Compiler/Checking/CheckExpressions.fs#L5126
But so far idk what is the current way.

Comment (abonie) [reply src/Compiler/Checking/CheckComputationExpressions.fs:2221]: We are essentially checking the same SynExpr (`comp`) twice, second time with a subtype constraint (first time to check if its type is `unit`)

**Comment by abonie** [comment]
Weird that it failed CI after rebasing on main with errors related to deserialization
```
Fatal error. Internal CLR error. (0x80131506)
at System.Runtime.Serialization.SerializationGuard.<ThrowIfDeserializationInProgress>g__ThrowIfDeserializationInProgress|0_0(System.Runtime.Serialization.SerializationInfo, System.String, Int32 ByRef)
```
Will try to rerun

**Review by edgarfgp — APPROVED** [review]

**Review by T-Gro — APPROVED** [review]

**Review by psfinaki — APPROVED** [review]

**Pull request merged**